### ISSUE TITLE
SmartOS support

### ISSUE THREAD

**Issue by rkh** [travis-ci/travis-ci#1368]
As an alternative to Linux and OSX. I have never used SmartOS, so no idea how hard this would be. I guess we could use Solaris containers on a SmartOS host system.

**Comment by roidrage** [comment]
Closing this issue for now, as we have no immediate plans to add this feature.

Should we add it to the roadmap eventually, we'll make sure to update this ticket.

**Comment by kigster** [comment]
That's too bad :(  SmartOS (while possible niche) is actually super nice for running server applications. 

Not intending to start an OS war here, but just to shed some light on why this "niche" OS is used by companies to run highly scalable and reliable ruby apps.  

We (at Wanelo) are hosted at Joyent Public Cloud (JPC), and when I was at ModCloth we also migrated the entire app toe Joyent without any hitch. Here are the five reasons I think SmartOS (and JPC) is awesome for deploying custom (ruby) applications:

* SmartOS used lightweight containers for many years before people discovered that software-based virtualization actually works pretty well.  You can (and could for years) dynamically resize your instances while they are running and continuing to serve requests. You can add/subtract more RAM and/or CPU cores, or disk space.

* SmartOS comes with a built-in service management framework called [SMF](http://www.cuddletech.com/blog/pivot/entry.php?id=182). SMF is bulletproof at managing and watching over services, and their spawned child processes if needed, restarting the master, etc. 

* SmartOS is a dream OS for performance geeks (like myself). It comes with dTrace, and everything you install compiles with dTrace enabled. This allows us to trace things like PostgreSQL's queries: their rate per second, average timing, slowest queries, etc. We also instrument our own ruby code with dTrace probes [supported by ruby 2.0](http://crypt.codemancers.com/posts/2013-04-16-profile-ruby-apps-dtrace-part1/). One of the best resources on this, [recently published book](http://www.amazon.com/gp/product/0133390098)

* SmartOS typically runs on ZFS, which comes with it's own amazing features, like copy on write, and file system snapshots and rollbacks. We are using this extensively, eg. for databases. By taking a snapshot every hour on the DB host, you have a solution when someone, say by accident, truncates or drops production tables. Without snapshots you have to run a replica with a delay, but unlike snapshots you do not get multiple time periods for recovery – you get as many as the number of replicas you've setup.  

* SmartOS has a concept of a "project" which is a custom defined aggregation of processes. Our [chef cookbooks](https://github.com/wanelo-chef) often define a project for a set of related services, for example we group various related sidekiq worker processes into a single project.  Majority of SmartOS system observation tools. like [prstat](http://www.brendangregg.com/DTrace/prstatvstop.html), have CLI flags that show by-project breakdown. Here is an example breakdown (taken from production, reproduced with fake project names for confidentiality): 

```
PROJID    NPROC  SWAP   RSS MEMORY      TIME  CPU PROJECT
   113       13 5243M 4642M   3.5%  23:47:25 2.9% sidekiq-email
   124        4 2036M 1749M   1.3%  25:15:55 2.8% sidekiq-feeds
     0       36  569M  425M   0.3% 1377:06:2 2.6% system
   104        4 1481M 1347M   1.0%   7:26:22 1.6% sidekiq-database-updates
   106        1  315M  287M   0.2%   5:03:26 0.7% sidekiq-counter-caches
   118        1  322M  292M   0.2%   4:44:01 0.7% sidekiq-cache-expiration
   116        2  651M  585M   0.4%   3:32:04 0.6% sidekiq-solr
   126        1  324M  293M   0.2%   3:18:51 0.6% sidekiq-es-saves
   108        3 1219M 1082M   0.8%   3:50:08 0.5% sidekiq-default
   105        1  322M  293M   0.2%   2:49:14 0.4% sidekiq-solr
```

You can see that this view shows how many processes, and independent threads are inside each project, and that information is invaluable in debugging performance problems. For example, we now have a well defined process for identifying how many threads vs processes to configure each group of sidekiq workers based on their workload as shown above. It uses `prstat -mLc -J` which gives microstate accounting, broken down by a project.

TL;DR

Once again, not here to start an OS war, as many other options, such as FreeBSD and Linux, are also valid, and have their own benefits, where a much wider adoption is probably the biggest one. That said, I don't see many resources published on exactly why SmartOS is used by some folks, so this is my take on it, after running large scale (ie 50K-200K RPMs) ruby applications for two companies over the last 6 years.

**Comment by dago** [comment]
I'd be interested in support for third party adapters too. It would allow using Travis as industry standard for non-mainstream operating systems and allow to broadening the test coverage e.g. to big-endian systems. At OpenCSW we are providing packages for Solaris and it would be nice to benefit directly from Travis integration. We already use the Webhooks from Github to trigger builds on Solaris and feed the status back for commits and pull requests.

Best regards -- Dago
